### PR TITLE
HPCC-18603 Add ESP thread to clean session

### PR DESCRIPTION
**Pull request by wangkx**
Also fix a bug for reading session timeout setting

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by wangkx** [comment]
@RussWhitehead please review.

File: esp/platform/espcfg.cpp [hunk 1]
```diff
@@ -144,6 +213,14 @@ void CEspConfig::ensureSDSSessionDomains()
             throw MakeStringException(-1, "Failed to connect to %s.", PathSessionRoot);
 
         ensureESPSessionInTree(conn->queryRoot(), m_process.str());
+
+        if (serverSessionTimeoutSeconds > ESP_SESSION_NEVER_TIMEOUT)
```
Comment (RussWhitehead) [inline esp/platform/espcfg.cpp]: I think "serverSessionTimeoutSeconds != ESP_SESSION_NEVER_TIMEOUT" is safer than presuming ESP_SESSION_NEVER_TIMEOUT is a negative number

File: esp/platform/espcfg.ipp [hunk 2]
```diff
@@ -101,12 +101,35 @@ struct esp_option
...
+    Semaphore  sem;
+
+    StringAttr espSessionSDSPath;
+    int        checkSessionTimeoutSeconds; //the duration to clean timed out sesssions
+
+public:
+    CSessionCleaner(const char* _espSessionSDSPath, int _checkSessionTimeoutSeconds) : Thread("CSessionCleaner"),
+        espSessionSDSPath(_espSessionSDSPath), checkSessionTimeoutSeconds(_checkSessionTimeoutSeconds) { }
+
+    ~CSessionCleaner()
```
Comment (RussWhitehead) [inline esp/platform/espcfg.ipp]: did you consider making this virtual?

File: esp/platform/espcfg.ipp [hunk 3]
```diff
@@ -101,12 +101,35 @@ struct esp_option
...
+    virtual int run();
+};
+
 class CEspConfig : public CInterface, implements IInterface 
 {
 private:
     Owned<IPropertyTree> m_envpt;
     Owned<IPropertyTree> m_cfg;
     Owned<IProperties> m_inputs;
+    Owned<CSessionCleaner> sessionCleaner;
```
Comment (RussWhitehead) [inline esp/platform/espcfg.ipp]: m_sessionCleaner

File: esp/platform/espcfg.cpp [hunk 4]
```diff
@@ -111,11 +111,62 @@ StringBuffer &CVSBuildToEspVersion(char const * tag, StringBuffer & out)
...
 }
 
+int CSessionCleaner::run()
+{
+    try
+    {
+        PROGLOG("CSessionCleaner Thread started.");
+
+        VStringBuffer xpath("%s*", PathSessionSession);
+        int checkSessionTimeoutMillSeconds = checkSessionTimeoutSeconds * 60;
```
Comment (RussWhitehead) [inline esp/platform/espcfg.cpp]: minor point, this variable isn't really necessary

**Review by RussWhitehead — COMMENTED** [review]
A couple trivial comments

File: esp/platform/espcfg.cpp (hunk 4 above)
Comment (wangkx) [reply esp/platform/espcfg.cpp]: But, I do not want to calculate it in every loops.

**Review by RussWhitehead — APPROVED** [review]
Looks good

**Comment by wangkx** [comment]
@richardkchapman this PR is ready to be merged.

**Pull request merged**